media/starboard/starboard_cdm.cc

```diff
@@ -12,371 +12,323 @@
-    const SessionDidNotUpdateCallback& session_did_not_update_callback) {
-  drm_system_->UpdateSession(*id_, key, key_length, session_updated_callback,
-                             session_did_not_update_callback);
+  NOTREACHED() << "Unexpected SbDrmSessionRequestType " << type;
 }
 
-void DrmSystem::Session::Close() {
-  drm_system_->CloseSession(*id_);
-  closed_ = true;
+CdmKeyInformation::KeyStatus ToCdmKeyStatus(SbDrmKeyStatus status) {
+  switch (status) {
+    case kSbDrmKeyStatusUsable:
+      return CdmKeyInformation::KeyStatus::USABLE;
+    case kSbDrmKeyStatusExpired:
+      return CdmKeyInformation::KeyStatus::EXPIRED;
+    case kSbDrmKeyStatusReleased:
+      return CdmKeyInformation::KeyStatus::RELEASED;
+    case kSbDrmKeyStatusRestricted:
+      return CdmKeyInformation::KeyStatus::OUTPUT_RESTRICTED;
+    case kSbDrmKeyStatusDownscaled:
+      return CdmKeyInformation::KeyStatus::OUTPUT_DOWNSCALED;
+    case kSbDrmKeyStatusPending:
+      return CdmKeyInformation::KeyStatus::KEY_STATUS_PENDING;
+    case kSbDrmKeyStatusError:
+      return CdmKeyInformation::KeyStatus::INTERNAL_ERROR;
+  }
+  NOTREACHED() << "Unexpected SbDrmKeyStatus " << status;
 }
 
-DrmSystem::DrmSystem(const char* key_system)
-    : wrapped_drm_system_(CreateSbDrmSystemWithHistogram(
-          key_system, this, OnSessionUpdateRequestGeneratedFunc,
-          OnSessionUpdatedFunc, OnSessionKeyStatusesChangedFunc,
-          OnServerCertificateUpdatedFunc, OnSessionClosedFunc,
-          media_metrics_provider_)),
+}  // namespace
+
+StarboardCdm::StarboardCdm(
+    const CdmConfig& cdm_config,
+    const SessionMessageCB& message_cb,
+    const SessionClosedCB& closed_cb,
+    const SessionKeysChangeCB& keys_change_cb,
+    const SessionExpirationUpdateCB& expiration_update_cb)
+    : message_cb_{message_cb},
+      closed_cb_{closed_cb},
+      keys_change_cb_{keys_change_cb},
+      expiration_update_cb_{expiration_update_cb},
       task_runner_(base::SequencedTaskRunner::GetCurrentDefault()),
-      ALLOW_THIS_IN_INITIALIZER_LIST(weak_ptr_factory_(this)),
-      weak_this_(weak_ptr_factory_.GetWeakPtr()) {
-  ON_INSTANCE_CREATED(DrmSystem);
-
-  if (is_valid()) {
-    LOG(INFO) << "Successfully created SbDrmSystem (" << wrapped_drm_system_
-              << "), key system: " << key_system << ".";
-  } else {
-    LOG(INFO) << "Failed to create SbDrmSystem, key system: " << key_system;
-  }
+      sb_drm_(SbDrmCreateSystem(cdm_config.key_system.c_str(),
+                                this,
+                                OnSessionUpdateRequestGeneratedFunc,
+                                OnSessionUpdatedFunc,
+                                OnSessionKeyStatusesChangedFunc,
+                                OnServerCertificateUpdatedFunc,
+                                OnSessionClosedFunc)) {
+  DCHECK(message_cb_);
+  DCHECK(closed_cb_);
+  DCHECK(keys_change_cb_);
+  DCHECK(expiration_update_cb_);
+
+  DCHECK(SbDrmSystemIsValid(sb_drm_));
```

Comment:
I was copying the approach we had in https://source.corp.google.com/h/lbshell-internal/cobalt_src/+/25.lts.1+:cobalt/media/base/drm_system.cc;l=82 where its constructed in the initializer list. I've added a check and LOG message to the constructor body. Does this work?

Reply:
At here https://source.corp.google.com/h/github/youtube/cobalt/+/25.lts.1+:cobalt/dom/eme/media_key_system_access.cc;l=53, the old implementation throws out an exception if it can't get a drm system. We can do something similar by adding a `bool is_valid()` function and check it in `StarboardCdmFactory`.

Reply:
thanks, added!